paddle/fluid/operators/ngraph/ngraph_engine.cc

```diff
@@ -0,0 +1,490 @@
+  ngraph::Shape sp;
+  for (int i = 0; i < dims.size(); ++i) {
+    int k = dims[i];
+    k = k == 0 ? 1 : k;
+    sp.push_back(k);
+  }
+  return sp;
+}
+
+static std::map<framework::proto::VarType::Type, ngraph::element::Type>
+    pd2ng_type_map = {
+        {framework::proto::VarType::FP32, ngraph::element::f32},
+        {framework::proto::VarType::FP64, ngraph::element::f64},
+        {framework::proto::VarType::INT32, ngraph::element::i32},
+        {framework::proto::VarType::INT64, ngraph::element::i64},
+        {framework::proto::VarType::BOOL, ngraph::element::boolean},
+};
+
+std::unordered_map<std::string, std::shared_ptr<ngraph::Function>>
+    NgraphEngine::func_cache_ = {};
+
+std::shared_ptr<ngraph::runtime::Backend> NgraphEngine::backend_ =
+    ngraph::runtime::Backend::create("CPU");
+
+static std::vector<std::vector<int>> NgraphOpIntervals(
+    framework::BlockDesc* block) {
+  std::vector<std::vector<int>> intervals;
+  auto ops = block->AllOps();
+  int size = ops.size();
+  int left = 0;
+  while (left < size && ops.at(left)->Type() != framework::kFeedOpType) {
+    ++left;
+  }
+  if (left == size) {
+    return intervals;
+  }
+  while (left < size && ops.at(left)->Type() == framework::kFeedOpType) {
+    ++left;
+  }
+
+  int right = left;
+  while (right < size && ops.at(right)->Type() != framework::kFetchOpType) {
+    ++right;
+  }
+  if (right == size) {
+    return intervals;
+  }
+  if (left >= right) return intervals;
+
+  // (left, right - 1) represents indices between feed and fetch
+  int pivot = left;
+  while (pivot < right) {
+    auto op_type = ops.at(pivot)->Type();
+    if (paddle::framework::NgraphBridge::NG_NODE_MAP.find(op_type) ==
+        paddle::framework::NgraphBridge::NG_NODE_MAP.end()) {
+      ++pivot;
+    } else {
+      int start = pivot, end = start;
+      while (pivot < right &&
+             (paddle::framework::NgraphBridge::NG_NODE_MAP.find(
+                  ops.at(pivot)->Type()) !=
+              paddle::framework::NgraphBridge::NG_NODE_MAP.end())) {
+        ++pivot;
+        ++end;
+      }
+      std::vector<int> interval = {start, end};
+      intervals.push_back(interval);
+    }
+  }  // end while
+  return intervals;
+}
+
+static void SubstituteNgraphOp(framework::BlockDesc* block,
+                               std::string block_str,
+                               std::vector<int> interval) {
+  framework::ProgramDesc program;
+  block->RemoveOp(interval.at(0), interval.at(1));
+  auto* ng_op = block->InsertOp(interval.at(0));
+  ng_op->SetType("ngraph_engine");
+  ng_op->SetAttr("interval", interval);
+  ng_op->SetAttr("graph", block_str);
+}
+
+void NgraphEngine::EnableNgraph(const framework::ProgramDesc& program) {
+#ifdef PADDLE_WITH_NGRAPH
+  VLOG(4) << "use_ngraph=True";
+  for (size_t bid = 0; bid < program.Size(); ++bid) {
+    auto* block =
+        const_cast<framework::ProgramDesc&>(program).MutableBlock(bid);
```

Comment:
@panyx0718 Please check if this const_cast is OK?

This is similar to `EnableMKLDNN` https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/framework/executor.cc#L516 but it's more complicated than MKLDNN. It will change the op_type in `ProgramDesc`.

Reply:
Add a todo to move EnableXXX to Compile time?
See this PR: https://github.com/PaddlePaddle/Paddle/pull/15089
In the CompiledProgram._compile, call EnableXXX. This saves some time when doing EnableXXX in Run all the time

Reply:
@baojun-nervana so what's your next move? I saw https://github.com/PaddlePaddle/Paddle/pull/14805#issuecomment-454325709 this is adding new op?

Reply:
@tensor-tang Are you okay with the const_cast or I have to address it before moving forward? I understand the moving EnableNgraph to compile time is more a performance optimization rather than addressing the const_cast issue.
@luotao1 Please go ahead on merging other PRs. I would rather addressing conflict than blocking other PRs.

Reply:
We are OK with the const_cast, you can just add one comment to follow when compile time is ready.

Reply:
And make sure merge the last code to pass CI

Reply:
Yeap, we will follow up on moving EnabledNgraph to compile time.

Reply:
@tensor-tang merged and rebased to develop, the PR_CI will be blocked by the const_cast though.

Reply:
@panyx0718  Please check the const_cast here.

Reply:
I don't see TODO being added here.

Reply:
@baojun-nervana Please add explicit TODO in code here.